File: amw/src/main/java/asl_actions/send_feedback.java
package asl_actions;

import jason.NoValueException;
import jason.asSemantics.DefaultInternalAction;
import jason.asSemantics.TransitionSystem;
import jason.asSemantics.Unifier;
import jason.asSyntax.*;

import javax.mail.*;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.Properties;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

public class send_feedback extends DefaultInternalAction {

	private static final String ORDER_CONFIRM = "Order confirm";
	private static final String ORDER_READY = "Order ready";
	private static final String ORDER_FAIL = "Order fail";
	private static final String ITEM_NOT_FOUND = "At least an item of your order has not been found";
	private static final String ITEM_RESERVED = "Due to another buy, at least one item of your order is not present in sufficient quantity";
	private static final BiFunction<String, List<String>, String> ORDER_INFO = ( orderId, items ) -> "Your order with id: " + orderId + " including the following items:\n\n"
			+ String.join( "\n", items ) + "\n\n has been confirmed";

	@Override
	public Object execute( TransitionSystem ts, Unifier un, Term[] args) throws NoValueException, IOException {

		String mailTo = args[ 0 ].toString( );
		String mailSubject, mailMsg;
		switch ( ( int ) ( ( NumberTerm ) args[ 1 ] ).solve( ) ) {
			case 200:
				mailSubject = ORDER_READY;
				mailMsg = ORDER_INFO.apply( args[ 2 ].toString( ), ( ( ListTerm ) args[ 3 ] ).stream( )
						.map( Term::toString ).collect( Collectors.toList( ) ) );
				break;
			case 202:
				mailSubject = ORDER_CONFIRM;
				mailMsg = ORDER_INFO.apply( args[ 2 ].toString( ), ( ( ListTerm ) args[ 3 ] ).stream( )
						.map( Term::toString ).collect( Collectors.toList( ) ) );
				break;
			case 404:
				mailSubject = ORDER_FAIL;
				mailMsg = ITEM_NOT_FOUND;
				break;
			case 409:
				mailSubject = ORDER_FAIL;
				mailMsg = ITEM_RESERVED;
				break;
			default:
				mailSubject = "";
				mailMsg = "";
		}

		BufferedReader reader = new BufferedReader( new FileReader( "testmaillogin" ) );

		Properties properties = new Properties();

		// Setup mail server
		properties.setProperty( "mail.host", "smtp.gmail.com" );
		properties.setProperty( "mail.smtp.port", "587" );
		properties.setProperty( "mail.smtp.auth", "true" );
		properties.setProperty( "mail.smtp.starttls.enable", "true" );

		String mailFrom = reader.readLine( );
		String pass = reader.readLine( );

		Session session = Session.getInstance( properties, new javax.mail.Authenticator( ) {
			protected PasswordAuthentication getPasswordAuthentication( ) {
				return new PasswordAuthentication( mailFrom, pass );
			}
		} );

		try {
			// Create a default MimeMessage object.
			MimeMessage message = new MimeMessage( session );

			// Set From: header field of the header.
			message.setFrom( new InternetAddress( mailFrom ) );

			// Set To: header field of the header.
			message.addRecipient( Message.RecipientType.TO, new InternetAddress( mailTo ) );

			// Set Subject: header field
			message.setSubject( mailSubject );

			// Now set the actual message
			message.setText( mailMsg );

			// Send message
			Transport.send( message );
		} catch ( MessagingException mex ) {
			mex.printStackTrace( );
			return false;
		}

		return true;
	}

}

File: src/main/kotlin/asl/action/send_feedback.kt
package asl.action

import asl.action.util.GoogleMailSender
import com.google.api.services.gmail.Gmail
import com.google.common.io.Resources
import jason.asSemantics.DefaultInternalAction
import jason.asSemantics.TransitionSystem
import jason.asSemantics.Unifier
import jason.asSyntax.ListTerm
import jason.asSyntax.NumberTerm
import jason.asSyntax.Term
import java.io.File
import java.util.*
import java.util.function.BiFunction
import javax.mail.*
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage


class send_feedback : DefaultInternalAction() {
	override fun execute(ts: TransitionSystem?, un: Unifier?, args: Array<Term>): Any {
		val mailTo = args[0].toString()
		val mailSubject: String
		val mailMsg: String

		TODO("https://blog.mailtrap.io/send-emails-with-gmail-api/")

		/*when ((args[1] as NumberTerm).solve().toInt()) {
			200 -> {
				mailSubject = ORDER_READY
				mailMsg = ORDER_INFO.apply(
					args[2].toString(),
					(args[3] as ListTerm).map { obj: Term -> obj.toString() }
				)
			}
			202 -> {
				mailSubject = ORDER_CONFIRM
				mailMsg = ORDER_INFO.apply(
					args[2].toString(),
					(args[3] as ListTerm).map { obj: Term -> obj.toString() }
				)
			}
			404 -> {
				mailSubject = ORDER_FAIL
				mailMsg = ITEM_NOT_FOUND
			}
			409 -> {
				mailSubject = ORDER_FAIL
				mailMsg = ITEM_RESERVED
			}
			else -> { mailSubject = ""; mailMsg = "" }
		}
		val reader = File(Resources.getResource("amw_mail_login").file).readLines()
		val properties = Properties()

		// Setup mail server
		properties.setProperty("mail.host", "smtp.gmail.com")
		properties.setProperty("mail.smtp.port", "587")
		properties.setProperty("mail.smtp.auth", "true")
		properties.setProperty("mail.smtp.starttls.enable", "true")
		//properties.setProperty( "mail.smtp.timeout", "5000" );

		val mailFrom = reader[0]
		val pass = reader[1]
		val session = Session.getInstance(properties, object : Authenticator() {
				override fun getPasswordAuthentication() = PasswordAuthentication(mailFrom, pass)
			})
		//session.setDebug( true );
		try {
			val message = MimeMessage(session)						// Create a default MimeMessage object
			message.setFrom(InternetAddress(mailFrom))				// SetFrom: header field of the header
			message.addRecipient(Message.RecipientType.TO, InternetAddress(mailTo))	// SetTo: header field of the header
			message.subject = mailSubject							// SetSubject: header field
			message.setText(mailMsg)								// Now set the actual message
			Transport.send(message)									// Send message
		} catch (mex: MessagingException) {
			mex.printStackTrace()
			return false
		}
		return true*/
	}

	companion object {
		private const val ORDER_CONFIRM = "Order confirm"
		private const val ORDER_READY = "Order ready"
		private const val ORDER_FAIL = "Order fail"
		private const val ITEM_NOT_FOUND = "At least an item of your order has not been found"
		private const val ITEM_RESERVED = "Due to another buy, at least one item of your order is not present in " +
				"sufficient quantity"
		private val ORDER_INFO = BiFunction { orderId: String, items: List<String>? ->
				"""Your order with id: $orderId including the following items: ${java.lang.String.join("\n",
					items)} has been confirmed"""
			}
	}
}